File: java/lib/src/main/java/io/openweb3/xwebhook/Authentication.java
package io.openweb3.xwebhook;

import io.openweb3.xwebhook.exceptions.ApiException;
import io.openweb3.xwebhook.internal.api.AuthenticationApi;
import io.openweb3.xwebhook.models.AppPortalAccessIn;
import io.openweb3.xwebhook.models.AppPortalAccessOut;
import io.openweb3.xwebhook.models.DashboardAccessOut;

public final class Authentication {
	private final AuthenticationApi api;

	Authentication() {
		api = new AuthenticationApi();
	}

	public AppPortalAccessOut appPortalAccess(final String appId, final AppPortalAccessIn appPortalAccessIn) throws ApiException {
		return this.appPortalAccess(appId, appPortalAccessIn, new PostOptions());
	}

	public AppPortalAccessOut appPortalAccess(final String appId, final AppPortalAccessIn appPortalAccessIn, final PostOptions options) throws ApiException {
		try {
			return api.v1AuthenticationAppPortalAccess(appId, appPortalAccessIn, options.getIdempotencyKey());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public DashboardAccessOut dashboardAccess(final String appId) throws ApiException {
		return this.dashboardAccess(appId, new PostOptions());
	}

	public DashboardAccessOut dashboardAccess(final String appId, final PostOptions options) throws ApiException {
		try {
			return api.v1AuthenticationDashboardAccess(appId, options.getIdempotencyKey());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public void logout() throws ApiException {
		this.logout(new PostOptions());
	}

	public void logout(final PostOptions options) throws ApiException {
		try {
			api.v1AuthenticationLogout(options.getIdempotencyKey());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}
}

File: kotlin/lib/src/main/kotlin/Authentication.kt
package io.openweb3.xwebhook.kotlin

import io.openweb3.xwebhook.kotlin.exceptions.ApiException
import io.openweb3.xwebhook.kotlin.internal.apis.AuthenticationApi
import io.openweb3.xwebhook.kotlin.models.AppPortalAccessIn
import io.openweb3.xwebhook.kotlin.models.AppPortalAccessOut
import io.openweb3.xwebhook.kotlin.models.DashboardAccessOut

class Authentication internal constructor(token: String, options: XwebhookOptions) {
    val api = AuthenticationApi(options.serverUrl)

    init {
        api.accessToken = token
        api.userAgent = options.getUA()
        options.initialRetryDelayMillis?.let { api.initialRetryDelayMillis = it }
        options.numRetries?.let { api.numRetries = it }
    }

    suspend fun appPortalAccess(
        appId: String,
        appPortalAccessIn: AppPortalAccessIn,
        options: PostOptions = PostOptions(),
    ): AppPortalAccessOut {
        try {
            return api.v1AuthenticationAppPortalAccess(appId, appPortalAccessIn, options.idempotencyKey)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun dashboardAccess(
        appId: String,
        options: PostOptions = PostOptions(),
    ): DashboardAccessOut {
        try {
            return api.v1AuthenticationDashboardAccess(appId, options.idempotencyKey)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun logout(options: PostOptions = PostOptions()) {
        try {
            api.v1AuthenticationLogout(options.idempotencyKey)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }
}
